test-integration-docker-compatibility: TestContainerListWithFilter is flaky

https://github.com/containerd/nerdctl/actions/runs/6859698743/job/18670305657

```
=== RUN   TestContainerListWithFilter
    container_list_linux_test.go:94: container nerdctl-testcontainerlistwithfilterlistWithFilterA is now running
    container_list_linux_test.go:94: container nerdctl-testcontainerlistwithfilterlistWithFilterB is now running
    container_list_linux_test.go:319: assertion failed: error is not nil: unexpected container nerdctl-testcontainerlistwithfilterlistWithFilterC found: CONTAINER ID   IMAGE                                       COMMAND              CREATED                  STATUS                  PORTS     NAMES
        5f3a56e64e6e   ghcr.io/stargz-containers/alpine:3.13-org   "top --restart=no"   Less than a second ago   Up Less than a second             nerdctl-testcontainerlistwithfilterlistWithFilterC
        5ee418d23fdb   ghcr.io/stargz-containers/alpine:3.13-org   "top"                1 second ago             Up Less than a second             nerdctl-testcontainerlistwithfilterlistWithFilterB
        20601c18d347   ghcr.io/stargz-containers/alpine:3.13-org   "top"                2 seconds ago            Up 1 second                       nerdctl-testcontainerlistwithfilterlistWithFilterA
        
--- FAIL: TestContainerListWithFilter (3.38s)
```

https://github.com/containerd/nerdctl/blob/462355d21065640514261872c81599757e6ff4a8/cmd/nerdctl/container_list_linux_test.go#L244-L574

--- a/cmd/nerdctl/container_list_linux_test.go
+++ b/cmd/nerdctl/container_list_linux_test.go
@@ -37,7 +37,8 @@ type psTestContainer struct {
 	network string
 }
 
-func preparePsTestContainer(t *testing.T, identity string, restart bool) (*testutil.Base, psTestContainer) {
+// When keepAlive is false, the container will exit immediately with status 1.
+func preparePsTestContainer(t *testing.T, identity string, keepAlive bool) (*testutil.Base, psTestContainer) {
 	base := testutil.NewBase(t)
 
 	base.Cmd("pull", testutil.CommonImage).AssertOK()
@@ -82,21 +83,23 @@ func preparePsTestContainer(t *testing.T, identity string, restart bool) (*testu
 		"-v", mnt1,
 		"-v", mnt2,
 		"--net", testContainerName,
-		testutil.CommonImage,
-		"top",
 	}
-	if !restart {
-		args = append(args, "--restart=no")
+	if keepAlive {
+		args = append(args, testutil.CommonImage, "top")
+	} else {
+		args = append(args, "--restart=no", testutil.CommonImage, "false")
 	}
 
 	base.Cmd(args...).AssertOK()
-	if restart {
+	if keepAlive {
 		base.EnsureContainerStarted(testContainerName)
+	} else {
+		base.EnsureContainerExited(testContainerName, 1)
 	}
 
 	// dd if=/dev/zero of=test_file bs=1M count=25
 	// let the container occupy 25MiB space.
-	if restart {
+	if keepAlive {
 		base.Cmd("exec", testContainerName, "dd", "if=/dev/zero", "of=/test_file", "bs=1M", "count=25").AssertOK()
 	}
 	volumes := []string{}

--- a/pkg/testutil/testutil.go
+++ b/pkg/testutil/testutil.go
@@ -316,6 +316,30 @@ func (b *Base) EnsureContainerStarted(con string) {
 	b.T.Fatalf("conainer %s not running", con)
 }
 
+func (b *Base) EnsureContainerExited(con string, expectedExitCode int) {
+	b.T.Helper()
+
+	const (
+		maxRetry = 5
+		sleep    = time.Second
+	)
+	var c dockercompat.Container
+	for i := 0; i < maxRetry; i++ {
+		c = b.InspectContainer(con)
+		if c.State.Status == "exited" {
+			b.T.Logf("container %s have exited with status %d", con, c.State.ExitCode)
+			if c.State.ExitCode == expectedExitCode {
+				return
+			}
+			break
+		}
+		b.T.Logf("(retry=%d)", i+1)
+		time.Sleep(sleep)
+	}
+	b.T.Fatalf("expected conainer %s to have exited with code %d, got status %+v",
+		con, expectedExitCode, c.State)
+}
+
 type Cmd struct {
 	icmd.Cmd
 	*Base

